SVA/lib/mmu.c

```diff
@@ -1379,6 +1312,160 @@ unmapSecurePage (struct SVAThread * threadp, unsigned char * v) {
   return;
 }
 
+/*   Function: ghostmemCOW()
+ *   
+ *   Description: 
+ *   Copy the parent's page table of ghost memory to the child. 
+ *   Write protect these page table entries for both the parent and the child.
+ *
+ *   Inputs:
+ *   oldThread - the SVAThread variable of the parent process
+ *   newThread - the SVAThread variable of the child process   
+ */
+void
+ghostmemCOW(struct SVAThread* oldThread, struct SVAThread* newThread)
+{
+    
+    uintptr_t vaddr_start, vaddr_end, size;
+    
+    vaddr_start = (uintptr_t) SECMEMSTART;
+    size = oldThread->secmemSize;
+    vaddr_end = vaddr_start + size;    
+
+   /*
+    * Get the PML4E of the new process's page table.  If there isn't one in the
+    * table, add one.
+    */
+    sva_integer_state_t integerState = newThread->integerState;
+    pml4e_t * pml4e =  (pml4e_t *) getVirtual(integerState.cr3 + secmemOffset);
+   
+    unprotect_paging(); 
+    if (!isPresent (pml4e)) {
+    /* Page table page index */
```

Comment:
Please be sure to indent code within blocks.  It looks like the code within the "if" clause isn't indented as it should be.

Please also be sure to indent using two spaces instead of using tabs and to wrap lines at 80 columns.  We are (kind of) following the LLVM Coding Standards in SVA.